File: 0576 Out of Boundary Paths.py
# Adapted from lee215
class Solution:
    def findPaths(self, m: int, n: int, N: int, i: int, j: int) -> int:
        res = [[0 for b in range(n)] for a in range(m)]
        for _ in range(N):
            res = [[(a == 0 or res[a-1][b]) + (a+1 == m or res[a+1][b]) \
                   + (b == 0 or res[a][b-1]) + (b+1 == n or res[a][b+1]) \
                   for b in range(n)] for a in range(m)]
            
        return res[i][j] % (10 ** 9 + 7)

        
'''
# Will TLE
class Solution:    
    def findPaths(self, m: int, n: int, N: int, i: int, j: int) -> int:
        def dfs(i, j, N):
            res = 0
            if N != 0:
                res += 1 if i - 1 < 0 else dfs(i-1, j, N-1)
                res += 1 if i + 1 == m else dfs(i+1, j, N-1)
                res += 1 if j - 1 < 0 else dfs(i, j-1, N-1)
                res += 1 if j + 1 == n else dfs(i, j+1, N-1)
            return res
        res = dfs(i, j, N)
        return res % (10 ** 9 + 7)

# Use dictionary to prevent TLE
class Solution:    
    def findPaths(self, m: int, n: int, N: int, i: int, j: int) -> int:        
        def dfs(i, j, N, dic):
            if (i, j, N) in dic:
                return dic[(i, j, N)]
            if N != 0:
                r = 0
                r += 1 if i - 1 < 0 else dfs(i-1, j, N-1, dic)
                r += 1 if i + 1 == m else dfs(i+1, j, N-1, dic)
                r += 1 if j - 1 < 0 else dfs(i, j-1, N-1, dic)
                r += 1 if j + 1 == n else dfs(i, j+1, N-1, dic)
                dic[(i, j, N)] = r
                return r
                
            return 0

        return dfs(i, j, N, {}) % (10 ** 9 + 7)

# LeetCode solution
class Solution:
    
    def findPaths(self, m: int, n: int, N: int, i: int, j: int) -> int:
        ret = self.findPathsHelper(m, n, N, i, j, {})
        return ret
    
    def findPathsHelper(self, m: int, n: int, N: int, i: int, j: int, memo: dict) -> int:
        
        if (i,j,N) in memo:
            return memo[(i,j,N)]
        
        if i < 0 or i >= m or j < 0 or j >= n:
            return 1
        
        if i >= N and i < m - N and j >= N and j < n - N:
            return 0
        
        if i >= m/2:
            return self.findPathsHelper(m, n, N, m-i-1, j, memo)
        
        if j >= n/2:
            return self.findPathsHelper(m, n, N, i, n-j-1, memo)
        
        ret1 = self.findPathsHelper(m, n, N-1, i-1, j, memo)
        ret2 = self.findPathsHelper(m, n, N-1, i, j-1, memo)
        ret3 = self.findPathsHelper(m, n, N-1, i+1, j, memo)
        ret4 = self.findPathsHelper(m, n, N-1, i, j+1, memo)
        
        ret = (ret1+ret2+ret3+ret4)%(10**9+7)
        memo[(i, j, N)] = ret
        
        return ret
'''

File: 0576 Out of Boundary Paths.java
// Adapted from shawngao
class Solution {
    public int findPaths(int m, int n, int N, int i, int j) {
        if (N <= 0) return 0;
        final int MOD = 1000000007;
        int[][] cnt = new int[m][n];
        cnt[i][j] = 1;
        int res = 0;
        int[][] dirs = {{-1, 0}, {1, 0}, {0, -1}, {0, 1}};
            
        for (int s = 0; s < N; ++s) {
            int[][] tmp = new int[m][n];
            for (int r = 0; r < m; ++r) {
                for (int c = 0; c < n; ++c) {
                    for (int[] d : dirs) {
                        int nr = r + d[0];
                        int nc = c + d[1];
                        if (nr < 0 || nr >= m || nc < 0 || nc >= n)
                            res = (res + cnt[r][c]) % MOD;
                        else
                            tmp[nr][nc] = (tmp[nr][nc] + cnt[r][c]) % MOD;
                    }
                }
            }
            cnt = tmp;
        }
        return res;
    }
}

/* Fatest solution at leetcode
class Solution {
  int M = 1000000007;

  public int findPaths(int m, int n, int N, int i, int j) {
    int[][][] memo = new int[m][n][N + 1];
    for (int[][] l : memo) for (int[] sl : l) Arrays.fill(sl, -1);
    return findPaths(m, n, N, i, j, memo);
  }

  public int findPaths(int m, int n, int N, int i, int j, int[][][] memo) {
    if (i == m || j == n || i < 0 || j < 0) return 1;
    if (N == 0) return 0;
    if (memo[i][j][N] >= 0) return memo[i][j][N];
    memo[i][j][N] = (
        (findPaths(m, n, N - 1, i - 1, j, memo) + findPaths(m, n, N - 1, i + 1, j, memo)) % M +
        (findPaths(m, n, N - 1, i, j - 1, memo) + findPaths(m, n, N - 1, i, j + 1, memo)) % M
    ) % M;
    return memo[i][j][N];
  }
}
*/